palette/src/js/script.js

```diff
@@ -0,0 +1,409 @@
+      document.getElementById('inp-color').value = this.currentColor;
+    } else if (idData === 'red-color') {
+      if (this.currentColor !== '#ff0000') {
+        this.prevColor = this.currentColor;
+        this.currentColor = '#ff0000';
+        document.querySelector('#current-color i').style.color = this.currentColor;
+        document.querySelector('#prev-color i').style.color = this.prevColor;
+        this.ctx.fillStyle = this.currentColor;
+        document.getElementById('inp-color').value = this.currentColor;
+      }
+    } else if (idData === 'blue-color') {
+      if (this.currentColor !== '#0000ff') {
+        this.prevColor = this.currentColor;
+        this.currentColor = '#0000ff';
+        document.querySelector('#current-color i').style.color = this.currentColor;
+        document.querySelector('#prev-color i').style.color = this.prevColor;
+        this.ctx.fillStyle = this.currentColor;
+        document.getElementById('inp-color').value = this.currentColor;
+      }
+    }
+  }
+
+  // finding coords depending on scale
+  getCoords(data) {
+    const box = this.canvas.getBoundingClientRect();
+    return {
+      top: Math.floor((data.pageY - box.top + window.pageYOffset) / (512 / this.scale)),
+      left: Math.floor((data.pageX - box.left + window.pageXOffset) / (512 / this.scale)),
+    };
+  }
+
+  // start drawing after mousedown
+  startDraw(data) {
+    this.permitDraw = true;
+    this.firstPoint.left = this.getCoords(data).left;
+    this.firstPoint.top = this.getCoords(data).top;
+    this.ctx.fillRect(
+      this.firstPoint.left,
+      this.firstPoint.top, 1, 1,
+    );
+  }
+
+  // stop drawing after mouseup or mouseout
+  stopDraw() {
+    this.permitDraw = false;
+    this.firstPoint.left = null;
+    this.firstPoint.top = null;
+  }
+
+  // drawing with Bresenham's line algorithm after mousemove
+  draw(data) {
+    if (this.permitDraw === true) {
+      let x1 = this.firstPoint.left;
+      let y1 = this.firstPoint.top;
+      const x2 = this.getCoords(data).left;
+      const y2 = this.getCoords(data).top;
+      this.firstPoint.left = this.getCoords(data).left;
+      this.firstPoint.top = this.getCoords(data).top;
+      const deltaX = Math.abs(x2 - x1);
+      const deltaY = Math.abs(y2 - y1);
+      const signX = x1 < x2 ? 1 : -1;
+      const signY = y1 < y2 ? 1 : -1;
+      let error = deltaX - deltaY;
+      this.ctx.fillRect(x2, y2, 1, 1);
+      while (x1 !== x2 || y1 !== y2) {
+        this.ctx.fillRect(x1, y1, 1, 1);
+        const error2 = error * 2;
+        if (error2 > -deltaY) {
+          error -= deltaY;
+          x1 += signX;
+        }
+        if (error2 < deltaX) {
+          error += deltaX;
+          y1 += signY;
+        }
+      }
+    }
+  }
+
+  fillArea(data) {
+    this.busy = true;
+    const startX = this.getCoords(data).left;
+    const startY = this.getCoords(data).top;
+    const fillColor = Palette.colorToRgba(this.currentColor);
```

Comment:
I think it would be better to store such static methods in some helper module. But I can be wrong